src/hashmap.cpp

```diff
@@ -0,0 +1,68 @@
+#include "hashmap.hpp"
+#include "csv.hpp"
+#include <sparsepp/spp.h>
+
+using spp::sparse_hash_map;
+
+Hashmap::Hashmap() {};
+
+Hashmap::Hashmap(const std::string &input_filename) {
+
+    io::CSVReader<3> in(input_filename);
+    in.set_header("from", "to", "speed");
+    external_nodeid_t from; external_nodeid_t to; congestion_speed_t speed;
+    // Pre-allocate a large chunk of memory to save on all the micro-allocations
+    // that would happen if adding items one-by-one and growing as needed
+    annotations.reserve(139064548);
+    while(in.read_row(from, to, speed)){
+        add(from, to, speed);
+    }
+
+};
+
+inline void Hashmap::add(const external_nodeid_t &from, const external_nodeid_t &to, const congestion_speed_t &speed) {
```

Comment:
Sometimes you pass ids by const ref, sometimes you pass them by value.

Because they're integers passing by (const) value is perfectly fine - can you adapt the code everywhere?